Conways logic

LucasUTNFRD opened the pull request:
## Changes included
This PR implements the core game logic for Conway's Game of Life, including:

1. Basic grid structure and visualization
2. Next generation calculation with neighbor counting
3. Project automation with Makefile
4. Documentation with README
5. Fixed neighbor boundary checking and offset calculations

LucasUTNFRD committed:
feat: Add next generation calculation with neighbor counting logic

- Implement next_generation() method to compute the next state of the grid
- Add count_neighbors() helper function to determine live neighbors for each cell

LucasUTNFRD committed:
feat: Add Makefile for project automation

- Add build, run, and test commands

LucasUTNFRD committed:
Add basic README

LucasUTNFRD committed:
feat: Add neighbors bound and offset checkings

LucasUTNFRD committed:
fix: Correct Conway's Game of Life rules implementation and add test coverage

The commit addresses an issue with pattern matching in the next_cell_generation
function and adds comprehensive test coverage for game logic verification.

Changes made:
- Fix pattern matching logic for Conway's rules implementation
- Add oscillator pattern tests (blinker, toad)
- Implement test cases to verify correct cell state transitions

dsocolobsky commented on README.md:
```diff
@@ -0,0 +1,3 @@
+# Conway's game of life
+Yet another Conway's game of life implementation
+
```
You should add instructions on how to clone the repo and run the code

LucasUTNFRD replied:
Will do it in next PR

dsocolobsky commented on TODO.md:
```diff
@@ -0,0 +1,13 @@
+## TODO
```
You could add the `TODO` directly in the `README.md` but it's fine if you want to keep it separate.

LucasUTNFRD replied:
Done

dsocolobsky commented on src/main.rs:
```diff
@@ -34,7 +35,9 @@ async fn main() {
+                    // This draws the cell
                     draw_rectangle(x as f32 * 10.0, y as f32 * 10.0, 10.0, 10.0, WHITE);
                 }
             }
```
This file is missing a blank line at the end; you should configure your editor to add it automatically.

lima-limon-inc replied:
This!
It's important for all files to have blank lines at the end. Some utilities depend on those to tell when a line ends!

LucasUTNFRD replied:
Will do it in next PR

dsocolobsky commented on src/conways.rs:
```diff
@@ -1,25 +1,263 @@
+        println!("Next generation:");
+        print_grid(&grid);
+    }
 }
```
This file is missing a blank line at the end; you should configure your editor to add it automatically.

LucasUTNFRD replied:
Will do it in next PR

dsocolobsky commented on Makefile:
```diff
@@ -0,0 +1,16 @@
+lint:
+	cargo clippy -- -D warnings
+
+.PHONY: all run test check lint
```
This file is missing a blank line at the end; you should configure your editor to add it automatically.

LucasUTNFRD replied:
Will do it in next PR

dsocolobsky reviewed (changes requested):
Code looks good to me, I left some minor style an README-related comments.

lima-limon-inc commented:
On it

damiramirez commented on src/conways.rs:
```diff
@@ -1,25 +1,263 @@
+                let ny = y as i32 + dy;
+
+                // Only count if within grid boundaries
+                if (nx >= 0 && nx < self.width as i32 && ny >= 0 && ny < self.height as i32)
+                    && self.grid[ny as usize][nx as usize] == CellState::Alive
```
Here you can use `get(idx)` and work with other `Option` functions like `and_then`, `is_some`, etc. With that, you don't need to check if the `idx` is outside the vector.

LucasUTNFRD replied:
Done!

damiramirez reviewed (approved):
LGTM + Dylan comments

lima-limon-inc commented on TODO.md:
```diff
@@ -0,0 +1,13 @@
+- [ ] Make clippy happy
+- [x] Add README
+- [x] Add Makefile 
+
+
+
```
Very minor nitpick, but it's better if files don't have extra newlines at the end.

1 is fine, and often required. More than that is a bit overkill

lima-limon-inc commented on src/conways.rs:
```diff
@@ -1,25 +1,263 @@
+                    (CellState::Alive, 2..=3) => CellState::Alive,
+                    // Rule 3: Any live cell with more than three live neighbors dies
+                    (CellState::Alive, 4..=8) => CellState::Dead,
+                    // Rule 4: Any dead cell with exactly three live neighbors becomes alive
```
Maybe you try doing a more generic "else" condition. 

If I'm not mistaken the following change:

```rust
                    (CellState::Alive, _) => CellState::Dead,
```

Will make it so that any case that wasn't caught with the previous two cases, will be caught here (as long as the state is alive).

This could be useful, specially if/when you change the size of the array to have more than 9 cells.

LucasUTNFRD replied:
In the next PR i will fix this

lima-limon-inc reviewed (approved):
Great job ^_^

LucasUTNFRD committed:
refactor: improve code readability and documentation

- Simplify boundary checking logic in Grid implementation
- Update README with project description and setup instructions
- Add newline at end of file for better git diff readability
